### ISSUE TITLE
Infer Types leading to `props: never`

### ISSUE THREAD

**Issue by lukebarton** [vercel/next.js#15913]
## Describe the bug
```ts
export const getServerSideProps = async ({ params }) => {
  return { 
    props: { foo: "bar" } 
  }
};

export const Page = (props: InferGetServerSidePropsType<typeof getServerSideProps>) => { ... }
```
causes `props: never`

however the following works fine:
```ts
export const getServerSideProps = async ({ params }: GetServerSidePropsContext) => {
  return { 
    props: { foo: "bar" } 
  }
};

export const Page = (props: InferGetServerSidePropsType<typeof getServerSideProps>) => { ... }
```
as does:
```ts
export const getServerSideProps = async (ctx) => {
  const params = ctx.params;
  return { 
    props: { foo: "bar" } 
  }
};

export const Page = (props: InferGetServerSidePropsType<typeof getServerSideProps>) => { ... }
```

**Comment by filipesmedeiros** [comment]
Next only checks [this](https://github.com/vercel/next.js/blob/canary/packages/next/types/index.d.ts#L128):

```typescript
export type InferGetServerSidePropsType<T> = T extends GetServerSideProps<
  infer P,
  any
>
  ? P
  : T extends (
      context?: GetServerSidePropsContext<any>
    ) => Promise<GetServerSidePropsResult<infer P>>
  ? P
  : never
```

So if neither the function is typed nor the argument, I think it has no way of knowing if the function in actually of type `GetServerSideProps`. I supose changing the above to something like: 

```typescript
export type InferGetServerSidePropsType<T> = T extends GetServerSideProps<
  infer P,
  any
>
  ? P
  : T extends (
      context?: any
    ) => Promise<GetServerSidePropsResult<infer P>>
  ? P
  : never
```

would make it infer your type, because the conditon would hold true?

**Comment by lukebarton** [comment]
I think you're right, however this one works, where the function isn't typed, and nor is the argument:

```ts
export const getServerSideProps = async (ctx) => {
```

**Comment by JuanM04** [comment]
It also happens in cases like this:

```ts
export const getServerSideProps = async (ctx) => {
  const { userId } = ctx.params
  const user = await getUser(userId)

  if(user) {
    return {
      props: { user }
    }
  } else {
    return { notFound: true }
  }
};
```

**Comment by tgallacher** [comment]
This also doesn't work:

```js
export const getServerSideProps: GetServerSideProps = async ({ params }) => {
  return { 
    props: { foo: "bar" } 
  }
};

export const Page = (props: InferGetServerSidePropsType<typeof getServerSideProps>) => { ... }
```

or 

```js
export const getServerSideProps: GetServerSideProps = async ({ params }: GetServerSidePropsContext) => {
  return { 
    props: { foo: "bar" } 
  }
};

export const Page = (props: InferGetServerSidePropsType<typeof getServerSideProps>) => { ... }
```

But removing the `GetServerSideProps` works:

```js
export const getServerSideProps = async ({ params }: GetServerSidePropsContext) => {
  return { 
    props: { foo: "bar" } 
  }
};

export const Page = (props: InferGetServerSidePropsType<typeof getServerSideProps>) => { ... }
```

**NextJS**: `v10.0.1`

For me, the magic was to make sure the `getServerSideProps` function args are explicitly typed using the `GetServerSidePropsContext` type. Otherwise, the page props are inferred as `any` type, e.g. `user: any`.

**Comment by GeeWee** [comment]
I can confirm that @tgallacher bottom sample also works for me.

**Comment by thisismydesign** [comment]
This is a sneaky one, my string props were mistyped but that didn't even cause a problem with strict TS. It only was an issue when I had an object in the props and wanted to access a property.

**Comment by nenadfilipovic** [comment]
This is how I solved it, maybe it will help someone

https://gist.github.com/nenadfilipovic/f2dd9cb903da93a7d14ed1de6b3493b1

**Comment by nenadfilipovic** [comment]
This is how I solved this, it will maybe help someone

https://gist.github.com/nenadfilipovic/f2dd9cb903da93a7d14ed1de6b3493b1#file-profile-tsx-L1-L32

**Comment by nenadfilipovic** [comment]
Here is my solution, key was to provide return type of getServerSideProps into GetServerSideProps type, here is example:
https://gist.github.com/nenadfilipovic/f2dd9cb903da93a7d14ed1de6b3493b1

**Comment by ezalivadnyi** [comment]
Setting types like this works:
```
GetServerSideProps<{
    user: User
    posts: Post[]
}>
```

**Comment by mkreuzmayr** [comment]
I found out something interessing on my machine.
Type inference via `InferGetServerSidePropsType<typeof getServerSideProps>` seems to be working as long as I only return props.

```
return { 
  props: { 
    user: { 
      firstName,
      lastName 
    } 
  } 
};
```

If i additionally return a redirect or a notFound conditionally type inference stops working for me.
```
return {
  redirect: {
    destination: '',
    permanent: false,
  },
};
```
```
return {
  notFound: true
};
```

**Comment by nenadfilipovic** [comment]
It probably works in first case because return type is
```
{ [key: string]: any }
```

**Comment by binajmen** [comment]
Having the same issue. Is there a proper solution? Not quite sure which one to choose in all the previous suggestions.

**Comment by mkreuzmayr** [comment]
@binajmen

The easiest solution imo would be by adding a PageProps type to `GetServerSideProps` generic. If there is a fix that the type infer starts working without the extra type specification you are already set up. And having your PageProps typed out is not a bad thing either.

```typescript
type PageProps = {
  user: { 
    firstName: string,
    lastName: string
  };
};

export const getServerSideProps: GetServerSideProps<PageProps> = async (ctx) => {
  return { 
    props: { 
      user: { 
        firstName,
        lastName 
      } 
    } 
  };
};

export const Page = ({ user }: InferGetServerSidePropsType<typeof getServerSideProps>) => { ... }
```

**Comment by KATT** [comment]
I've found a workaround - first of all - something broke with `11.1.x` as could use `InferGetServerSidePropsType<typeof getServerSideProps>` even with `notFound` using the below.


Here's my hacky implementation that I just wrote:



### `utils/inferSSRProps.ts`
```ts
/* eslint-disable @typescript-eslint/no-explicit-any */

type GetSSRResult<TProps> =
  //
  { props: TProps } | { redirect: any } | { notFound: true };

type GetSSRFn<TProps extends any> = (args: any) => Promise<GetSSRResult<TProps>>;

export type inferSSRProps<TFn extends GetSSRFn<any>> = TFn extends GetSSRFn<infer TProps>
  ? NonNullable<TProps>
  : never;
```


### `pages/somePage.tsx`

```tsx
import { inferSSRProps } from '../utils/inferSSRProps'
import { GetServerSidePropsContext } from "next";
import prisma from "@lib/prisma";

export default MyPage(props: inferSSRProps<typeof getServerSideProps>) {
  // ...
}

export const getServerSideProps = async (context: GetServerSidePropsContext) => {
  const post = await prisma.post.findFirst({
    where: {
      username: (context.query.slug as string).toLowerCase(),
    },
    select: {
      id: true,
      // ...
    },
  });
  if (!post) {
    return {
      notFound: true,
    } as const; // <-- important, this needs to be `as const`
  }
  return {
    props: {
     post,
    },
  };
}

**Comment by mkreuzmayr** [comment]
Hey @KATT, thanks for your solution!

You do not need to cast `{ notFound: true }` to `const` if you change your `GetSSRResult` `notFound` type to `boolean`.

### `utils/inferSSRProps.ts`

```typescript
/* eslint-disable @typescript-eslint/no-explicit-any */

type GetSSRResult<TProps> =
  { props: TProps } | { redirect: any } | { notFound: boolean }; // <-------

type GetSSRFn<TProps extends any> = (args: any) => Promise<GetSSRResult<TProps>>;

export type inferSSRProps<TFn extends GetSSRFn<any>> = TFn extends GetSSRFn<infer TProps>
  ? NonNullable<TProps>
  : never;
```

**Comment by balazsorban44** [comment]
Might be interesting https://github.com/microsoft/TypeScript/issues/38511

**Comment by Markyiptw** [comment]
> I found out something interessing. Type inference via `InferGetServerSidePropsType<typeof getServerSideProps>` seems to be working as long as I only return props.
> 
> ```ts
> return { 
>   props: { 
>     user: { 
>       firstName,
>       lastName 
>     } 
>   } 
> };
> ```
> 
> If i additionally return a redirect or a notFound conditionally type inference stops working for me.
> 
> ```ts
> return {
>   redirect: {
>     destination: '',
>     permanent: false,
>   },
> };
> ```
> 
> ```ts
> return {
>   notFound: true
> };
> ```
> 
> > Setting types like this works:
> > ```
> > GetServerSideProps<{
> >     user: User
> >     posts: Post[]
> > }>
> > ```
> 
> This also works for me, but it would be more comfortable if it would work without the extra type specification. If the type is created like this, you would not need to infer it, because you can directly use it in the component.

returning an empty props in redirect worked for me:

```ts
return {
  redirect: {
    destination: '',
    permanent: false,
  },
  props: {}
}
```

**Comment by aengl** [comment]
Thanks @Markyiptw for the hint, that helped me a lot in figuring out what was going on.

The disadvantage of that solution is that the props will all become optional, so that may not always be ideal.

I slapped together this custom type that seems to do work well when using `getServerSideProps` that may return a redirect:

```
export type CustomInferGetServerSidePropsType<T> = T extends (
  context?: any
) => Promise<{ props: infer P }>
  ? P
  : T extends (context?: any) => Promise<GetServerSidePropsResult<infer P>>
  ? P
  : never;
```

I'm sharing this because the other types shared in this issue didn't quite work for my cases.

**Comment by PabloLION** [comment]
I found a best work around.
remove the `: GetStaticProps` from `const getStaticProps = async () => {...}`

Then the `const SomePage: NextPage<InferGetStaticPropsType<typeof getStaticProps>>= ({/* inferred types */}){} ` can work correctly. 

Problem was as said here https://github.com/vercel/next.js/issues/32434#issuecomment-993013691. So all after adding the `: GetStaticProps` , the return type of getStaticProps would be extended to `{ [key: string]: any; }` by TS because it includes the original type of {'foo': string}.

**Comment by flybayer** [comment]
Looks like this will hopefully be solved with Typescript 4.9!

https://twitter.com/leeerob/status/1563540593003106306

<img width="606" alt="image" src="https://user-images.githubusercontent.com/8813276/187041520-21a37d39-bd27-489f-9305-42ca78dc2dee.png">

**Comment by andparsons** [comment]
I was playing around with the new satisfies keyword and I can certainly see it helping.

We’ll still need to cast the notFound and redirects returns as consts.

I’m also of the opinion that we should expect props to be able to return null unless a type guard is in place, which I’ve also included in the below example.

[Link to ts playground](https://www.typescriptlang.org/play?ts=4.9.0-dev.20220827#code/C4TwDgpgBACgTgezAZygXigOwK4BtdQA+UA3gFBRQD0AVDVAAIAmEYcEAxgIbARNQ0qFKAEsmALijJgcEZgDmAbjIBfZWQ4JM0qPIjAAysB4iO8JKgxdkITBygAzbHeAitUABSbMvAB7AASlJhbx0HfQ4ACwMEAFsIcxQAFUieAGEEPCYAIQgAOTwCDBJRCSgAIgBGACYAZnKoFShrWEQUdUoRB09w4CiY+MTkFPTM3Bz8wvQ0DBx8IPJKSlDgLARgADFMzH5i4SWodgA3LlwxHghJB1PkCAAafaXMda3nMplse8emlpXlA-YwGwcEwa022yY-0aZGEXR6EWicQSbWGqWAGSyuQK+AAdGJphhvH5gDiWNc8MAADIIbi4CALR4rQ58ETsDirPYHSjHU7nXhXG5fLnMpiszjASSLYVQFjSOQmLSSAAGVCVD2lUmMQOQGRYklqAAZKuquSoTZQfqg-o9AcDQexRWzgFCVDDufo7cEARATmcmBdJMbHmwLFcEQNkRYRuixhNsbgTWpVFITMgHCIIKgAOL6IwmMwo9QQXxgBBwVZkrgUxzOdluUEwLh6DzCEMoSQASUw4TgOcMWtMQyS4AgAB5QJAEN09P380MAHxkBmUNvIPGQ4RUKhQAB6AH4PGAuHAuPFeHAgqvJaVJNJZApFK7XWRaDQKPQhlJIrGoCGAEZcH+uAgM0uDIAgUB-tApy4AgADufBQMAEFQVghR3FIEEIQA5PgoGwXBjhlkhaKiKCwCRCIqC8NI75CMWpblkhI5QH2eauAWFijsIMBQMWvA7KgJQANoANYQCAt4yHI8gALqSFwmAgU0xBzEUpBQGJElSfeckKUpjREGh+AmgAinx-gQIJsDHrcTAAKpwLgpmfHAIEYI2cB2Y5zmuSAJoACIWQJTCoPAPoZnBAU8Fw6CtBFEBRTFZDznFLbLFoxKSGxA6cSgGQ+Pxo6mRhAWLkEaCpeYsRUWOOVziiABKmYUqOMDzqlxD1RxQzNcgrXteoL50FABhwVwYBgDJjiILEMqmKsyHMpocD8AAVtgYTERRVFIZmwAYc8qxyCFiEOMRUSKXoAhCBO0Ced5TkuRAblxc1K1MKOd4yRhziic8cGYPO6h3VAzWOuK6DCMQUoHNIPCbbqlxQIalRGYa1TowatRYwA7FjAAcUIHLKriYAqmA6TJxNLABtyNhRe4CmBEA066lAw48K4vbEilWRKkEIAgdKKTTlCk-KriKpquli5B1gJDwkRM44gps0WJZlotLHhUckXRcYcXfQoRnXCzRkIH+a2Q8QbwQOmmB8DCDFa8xkCsbmuVDAVxLcZQ5n8VZoU2V5fA+c9r0ebZYdPX5gXBUHYXHPrMVxbrKfGClcVSkeJ6xMgKuma2yeJSrf5CyLmDFwlSXGCrAXCLBtIQCrxvyI3NKnJmrfSQowmycIlYUtSzc97pqjO5rTGg91g5NS1uDAG1qVoNDGlXrAijMr6fIt5IOCxFBcBGeXwsQIp0IcxpDpiuykjg7fzrb7y-q8CrB9HyfFfn6C7NGSUR1Xg7EkB8Vmz8-QXHftgQ+L0v5nwvs+F208WJdh7LPPKwwRyjiSCvKASQE7WXQUMUccgeywAwopEAi5KB7lgMISQ+DA7WXSlyIk-EVZEJRD7IqlDqFLAqlVWatVRycIsH1VqpDYHtU6h7WcPV579UXiQ7sUi+G0JgPQrAPoXpkCAA)

**Comment by HaNdTriX** [comment]
I have created a PR to solve this issue^^

**Comment by HaNdTriX** [comment]
This issue has been fixed by #40635

**Comment by ijjk** [comment]
Closing per above
